vm/devices/storage/disk_backend/src/layered.rs

```diff
@@ -0,0 +1,718 @@
+                is_fua_respected &= layer.is_fua_respected;
+                let unmap = match layer.unmap_behavior {
+                    UnmapBehavior::Zeroes => true,
+                    UnmapBehavior::Unspecified => !unmap_must_zero,
+                    UnmapBehavior::Ignored => false,
+                };
+                if !unmap {
+                    optimal_unmap_sectors = None;
+                } else if let Some(n) = &mut optimal_unmap_sectors {
+                    *n = (*n).max(layer.optimal_unmap_sectors);
+                }
+            }
+            last_write_through = layer.behavior.write_through;
+            if disk_id.is_none() {
+                disk_id = layer.disk_id;
+            }
+        }
+        if last_write_through {
+            return Err(InvalidLayeredDisk::UselessWriteThrough {
+                layer: layers.len() - 1,
+            });
+        }
+
+        let sector_size = layers[0].sector_size;
+        let physical_sector_size = layers[0].physical_sector_size;
+
+        let mut visible_sector_count = !0;
+        let layers = layers
+            .into_iter()
+            .map(|layer| {
+                visible_sector_count = layer.sector_count.min(visible_sector_count);
+                Layer {
+                    behavior: layer.behavior,
+                    backing: layer.backing,
+                    visible_sector_count,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        Ok(Self {
+            is_fua_respected,
+            is_read_only,
+            sector_shift: sector_size.trailing_zeros(),
+            disk_id,
+            physical_sector_size,
+            optimal_unmap_sectors,
+            layers,
+        })
+    }
+}
+
+trait DynLayer: Send + Sync + Inspect {
+    fn sector_count(&self) -> u64;
+
+    fn read<'a>(
+        &'a self,
+        buffers: &'a RequestBuffers<'_>,
+        sector: u64,
+        bitmap: SectorMarker<'a>,
+    ) -> Pin<Box<dyn 'a + Future<Output = Result<(), DiskError>> + Send>>;
+
+    fn write<'a>(
+        &'a self,
+        buffers: &'a RequestBuffers<'_>,
+        sector: u64,
+        fua: bool,
+        no_overwrite: bool,
+    ) -> Pin<Box<dyn 'a + Future<Output = Result<(), DiskError>> + Send>>;
+
+    fn sync_cache(&self) -> Pin<Box<dyn '_ + Future<Output = Result<(), DiskError>> + Send>>;
+
+    fn unmap(
+        &self,
+        sector: u64,
+        count: u64,
+        block_level_only: bool,
+        next_is_zero: bool,
+    ) -> Pin<Box<dyn '_ + Future<Output = Result<(), DiskError>> + Send>>;
+
+    fn wait_resize(&self, sector_count: u64) -> Pin<Box<dyn '_ + Future<Output = u64> + Send>>;
+}
+
+impl<T: LayerIo> DynLayer for T {
+    fn sector_count(&self) -> u64 {
+        self.sector_count()
+    }
+
+    fn read<'a>(
+        &'a self,
+        buffers: &'a RequestBuffers<'_>,
+        sector: u64,
+        bitmap: SectorMarker<'a>,
+    ) -> Pin<Box<dyn 'a + Future<Output = Result<(), DiskError>> + Send>> {
```

Comment:
shouldn't these be using StackFuture?

Reply:
Ehh. We need to reevaluate `StackFuture` anyway, since a lot has changed since we started using it in OpenHCL.